Enhancement/9692 update checkbox design

abdelmalekkkkk opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #9692

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil reviewed (changes requested):
Thanks @abdelmalekkkkk, good work here. It does need a few amendments to more fully match the Figma design, please see my comments.

Additionally, I've realised it doesn't seem possible to fully match one aspect of the design - or at least, not without a substantial rework, as far as I can tell.

The hover state in Figma has a filled background for the checkbox, but it doesn't seem possible to the background for a checkbox. We'd need to hide the input and show our own version of it, and it's preferable to avoid going down that route.

Please make a note of this in the QAB to make it clear for QA engineers, potentially avoiding an unnecessary return to execution for the issue.

<img width="67" height="59" alt="image" src="https://github.com/user-attachments/assets/762d93f4-ed41-4169-8b47-91cde251e5b1" />

abdelmalekkkkk commented on assets/sass/vendor/_mdc-checkbox.scss:
```diff
@@ -31,6 +77,42 @@
+				}
+			}
+
+			&:enabled:not(:checked, :indeterminate),
```
I had to add that to override the existing CSS since it's a more specific selector.

techanvil replied:
Hmm, would you mind clarifying that further? In my testing, looking at the checkboxes in Storybook, the styles seem OK when I remove the `:indeterminate` pseudo-class from the selector. Not sure if I'm missing something.

techanvil commented on assets/js/googlesitekit/components-gm2/Checkbox.stories.js:
```diff
@@ -74,6 +88,16 @@ Disabled.args = {
+	id: 'googlesitekit-checkbox-9',
+	name: 'googlesitekit__checkbox-9',
+	value: 'value-9',
+	label: 'Disabled Checkbox',
```
Let's amend the label to distinguish it from the unchecked version of the disabled checkbox:

```suggestion
	label: 'Disabled Checked Checkbox',
```

techanvil commented on assets/sass/vendor/_mdc-checkbox.scss:
```diff
@@ -31,6 +82,42 @@
+
+			&[disabled]:checked:not(:indeterminate) {
+				~ .mdc-checkbox__background {
+					background-color: $c-interactive-on-disable-container;
```
Apologies, I realise I made a typo in my previous comment where I advised the correct colour token to use here. 

It should in fact be `$c-interactive-disable-container`.

```suggestion
					background-color: $c-interactive-disable-container;
```

<img width="1515" height="879" alt="image" src="https://github.com/user-attachments/assets/351f15c2-3d9d-4771-9956-f8cecf18d4fd" />

techanvil commented on assets/sass/vendor/_mdc-checkbox.scss:
```diff
@@ -31,6 +82,42 @@
+
+			&:enabled:not(:checked, :indeterminate) {
+				~ .mdc-checkbox__background {
+					background-color: transparent;
```
Is this background colour style definitely needed? Again, maybe I'm missing something, but it seems OK when I remove it...

techanvil replied:
Thanks for clarifying, sorry I totally missed what now seems quite obvious!

techanvil reviewed (changes requested):
Hey @abdelmalekkkkk, thanks for updating this PR. It does need a couple of further changes, please see my comments.

> I applied all the requested changes and also fixed your very first comment regarding the checkbox rectangle's filled background.

Also, please can you clarify what you fixed with regard to the filled background?

In my comment I was pointing out the difficulty of providing a fill for the checkbox background, but looking at Storybook for the latest version of the branch, the checkbox background is still white when hovering over it.

<img width="335" height="518" alt="image" src="https://github.com/user-attachments/assets/ad044127-f1d0-46e4-9492-513f5ecbc1b1" />

abdelmalekkkkk commented:
@techanvil I addressed your comments and answered your questions. Regarding the checkbox background, I have fixed it (again) to have a filled background instead of a white one. The problem was that the selector needed to have the simple `:not` notation (`:not(:checked):not(:indeterminate)`) instead of the complex one (`not(:checked, :indeterminate)`) to properly override the background.

Please take another look and let me know if all is good.

techanvil commented:
> @techanvil I addressed your comments and answered your questions.

Thanks @abdelmalekkkkk, that's great.

> Regarding the checkbox background, I have fixed it (again) to have a filled background instead of a white one. The problem was that the selector needed to have the simple `:not` notation (`:not(:checked):not(:indeterminate)`) instead of the complex one (`not(:checked, :indeterminate)`) to properly override the background.
> 
> Please take another look and let me know if all is good.

Right, I see. Sorry to be a bit slow on the uptake, but now I understand. Evidently the simple selector is more specific because it's specifying two `not` pseudo-classes as opposed to a single, combined `not`.

<img width="868" height="582" alt="image" src="https://github.com/user-attachments/assets/aaa31e36-7f16-49ec-b420-24e2da91444c" />

techanvil commented on assets/sass/vendor/_mdc-checkbox.scss:
```diff
@@ -31,6 +82,44 @@
+				}
+			}
+
+			/* stylelint-disable selector-not-notation -- The complex notation (not(:checked, :indeterminate))
+		    * doesn't seem to override the styles. */
```
Sorry, one last, nit-picky feedback comment and this is good to go :sweat_smile: 

Please can you update this comment to be a bit less vague? As per this [comment](https://github.com/google/site-kit-wp/pull/11138#issuecomment-3166296633), we can be clear about why the simple notation is needed here.

techanvil reviewed (commented):
Thanks @abdelmalekkkkk, that's great!

Btw, I noticed that an unrelated VRT reference image had been updated and its test was failing in CI. I've [reverted the change myself](https://github.com/google/site-kit-wp/pull/11138/commits/8c5d733ee0a67ec9bebfe2c04471df6a4379b0fc) to save another round trip.
